src/frontend/src/optimizer/plan_node/stream_upstream_sink_union.rs

```diff
@@ -0,0 +1,145 @@
+// Copyright 2025 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use educe::Educe;
+use pretty_xmlish::XmlNode;
+use risingwave_common::catalog::Schema;
+use risingwave_pb::stream_plan::UpstreamSinkUnionNode;
+use risingwave_pb::stream_plan::stream_node::PbNodeBody;
+
+use super::stream::prelude::*;
+use crate::OptimizerContextRef;
+use crate::expr::ExprImpl;
+use crate::optimizer::plan_node::expr_visitable::ExprVisitable;
+use crate::optimizer::plan_node::utils::{Distill, childless_record};
+use crate::optimizer::plan_node::{ExprRewritable, PlanBase, Stream, StreamNode};
+use crate::optimizer::property::{
+    Distribution, FunctionalDependencySet, MonotonicityDerivation, MonotonicityMap,
+    WatermarkColumns, analyze_monotonicity,
+};
+use crate::stream_fragmenter::BuildFragmentGraphState;
+
+#[derive(Debug, Clone, Educe)]
+#[educe(PartialEq, Eq, Hash)]
+struct UpstreamSinkUnionInner {
+    #[educe(PartialEq(ignore), Hash(ignore))]
+    ctx: OptimizerContextRef,
+    schema: Schema,
+    append_only: bool,
+    // `generated_column` is used to generate the `watermark_columns` field and affects the derivation of subsequent
+    // operators. Since the project operator of `generated_column` is actually on the sink-fragment, not on the table,
+    // only expr can be retained here to determine the non-decreasing column.
+    generated_column_exprs: Option<Vec<ExprImpl>>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct StreamUpstreamSinkUnion {
+    pub base: PlanBase<Stream>,
+    inner: UpstreamSinkUnionInner,
+}
+
+impl StreamUpstreamSinkUnion {
+    pub fn new(
+        ctx: OptimizerContextRef,
+        schema: &Schema,
+        append_only: bool,
+        generated_column_exprs: Option<Vec<ExprImpl>>,
+    ) -> Self {
+        let inner = UpstreamSinkUnionInner {
+            ctx,
+            schema: schema.clone(),
+            append_only,
+            generated_column_exprs,
+        };
+
+        Self::new_inner(inner)
+    }
+
+    fn new_inner(inner: UpstreamSinkUnionInner) -> Self {
+        let mut out_watermark_columns = WatermarkColumns::new();
```

Comment:
Can we by any chance to reuse the logic in `Project` to obtain the watermark columns?